src/librustc_ast_lowering/path.rs

```diff
@@ -55,7 +56,21 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                         Res::Def(DefKind::AssocTy, def_id) if i + 2 == proj_start => {
                             Some(parent_def_id(self, def_id))
                         }
-                        Res::Def(DefKind::Variant, def_id) if i + 1 == proj_start => {
+                        Res::Def(DefKind::Variant, def_id)
+                            if i + 2 == proj_start && segment.args.is_some() =>
+                        {
+                            // Handle `Enum::<Param>::Variant {}`.
+                            // We need to choose one here so that the method `res_to_ty` in
+                            // `astconv` works correctly, but it has to be *only* one, so track
+                            // whether we've already set the `DefId` in the previous segment.
+                            type_param_in_enum = true;
+                            Some(parent_def_id(self, def_id))
+                        }
+                        Res::Def(DefKind::Variant, def_id)
+                            if i + 1 == proj_start
+                                && (segment.args.is_some() || !type_param_in_enum) =>
+                        {
+                            // Handle `Enum::Variant::<Param> {}`.
                             Some(parent_def_id(self, def_id))
                         }
```

Comment:
Two things here:
1. you can avoid mutable state by looking at `p.segments`
  * btw I doubt a bit that `.args.is_some()` is the correct way to check this, it might be true even for `::<>` which I'm not sure if it's compatible with the similar check in typeck
2. these only make sense if the param mode is explicit, which it can never be for `Expr::Struct` right now (`<Enum>::Variant {...}` is not valid syntax, and `Enum::Variant` is not valid in type paths *yet*, only one of these two could actually be relevant here)
  * the comments referring to `Enum::Variant {...}` are therefore inaccurate IMO
  * this needs some extra checks, probably for the param mode being explicit
  * it's really mostly forward compat for `enum` variant types so you could just as well add an assert that `DefKind::Variant` parts are never `ParamMode::Explicit`